### PR TITLE
Merge Unix and Windows build systems in the root directory

### PR DESCRIPTION
**Pull request by shindere**
Cc @xavierleroy @damiendoligez @alainfrisch @dra27 @adrien-n

This PR merges the two makefiles (Makefile and Makefile.nt) at the root
of the OCaml repository.

Although this PR tries to introduce a few easy improvements in addition to
the merge of build systems itself, there are many things that remain
to be improved and, given the current size of this PR (more than 100
commits), it is probably better to delay these remaining improvements until
after this PR has been merged.

In particular, although this PR may fix things that did not work before,
no particular effort has been done in this direction. More
efforts were invested in not breaking anything that was working
before and not introducing any regression. So, it may be a good idea,
while reviewing and making comments, to focus on ensuring that
nothing got accidentally broken, rather
than on what could be
improved from there, at least for a first pass.

## Changes affecting both build systems

 * Introduce the BYTECODE_C_COMPILER and NATIVE_C_COMPILER make variables
   These variables represent the C compilers ocamlc and ocamlopt
   should use to compile a third-party C source file when no -cc
   command-line option has been specified.
 * Remove the PARTIALLD make variable since it is unused
 * Get rid of the ARCH_OCAMLOPT make variable
   (It is a synonymous of the ARCH variable and is not exported.)

## Changes specific to the Unix build system

 * Stop removing -Werror from compiler options
   There was a hack to ensure -Werror will not be passed to the C compiler
   when called by ocamlc/ocamlopt to compile third-party C source files.
   Given that the flags used to compile third-party C source files
   have now been distinguished from those used to compile those C
   files that belong to the compiler distribution, this hack is no
   longer necessary.
 * Stop removing shared libraries during install
   This had been introduced in commit a49cf7b0 but does not seem useful
   any longer.
 * ocamldep is now called with the -slash command-line option as is done
   on Windows. This should not have any effect, though.
 * The coldstart rule was creating a symlink to ../byterun/caml in the
   stdlib directory. Stop doing that because it was not done under
   Windows and it seems not creating this symlink does not break anything.
 * The Unix build system has a base.opt rule that this PR currently preserves,
   but since this rule is not used anywhere and has no counterpart on
   the Windows build system, its usefulness probably needs to be confirmed.

## Changes specific to the Windows build system

 * Stop installing Changes, README and LICENSE files
 * Take the configured value of the following variables into account when
   generating utils/config.ml from utils/config.mlp instead of using
   hard-coded values:
   SYSTHREAD_SUPPORT, ASM_CFI_SUPPORTED, LIBUNWIND_AVAILABLE,
   LIBUNWIND_LINK_FLAGS and CC_PROFILE. Also, define the
   WITH_FRAME_POINTERS in config/Makefile.m* and make use of this
   definition, instead of using a hard-coded value.
 * Make the install-compiler-sources target available
   It had been introduced by PR#827 but only on the Unix side
 * Fix the installopt rule to install:
 ** the .mli files from the middle_end, middle_end/base_types and
    asmcomp directories to $(INSTALL_COMPLIBDIR), as is done under Unix.
 ** ocamlc$(EXE) and ocamllex$(EXE) as copies of (or symlinks to)
    ocamlc.byte$(EXE) and ocamllex.byte$(EXE), when installopt is
    executed but installoptopt is not, as is done on Unix.
 * Install .mli files to the compilerlibs directory
   Before this PR, the mli files of the utils, parsing, typing,
   bytecomp, driver and toplevel directories were not installed to
   the compilerlibs directory on Windows as was done on Unix.
   This PR fixes this.
 * Install more topdirs-related files to INSTALL_LIBDIR
   Before this PR, only topdirs.cmi was installed to INSTALL_LIBDIR
   on Windows. This PR also installs topdirs.cmt, topdirs.cmti and
   topdirs.mli, as is done on Unix.
 * The manual-pregen and html_doc targets become available on Windows
   (they have not been tested yet, though.)
 * The checkstack target also becomes available on Winows
   (but checkstack itself does not work yet)
   The checkstack target has been made more verbose and the $(EXE)
   extension is now used in a more coherent way.
 * Add dependencies to the ocamllex and ocamllex.opt targets to match
   what is done on Unix. However the dependencies between ocamllex
   and ocamllex.opt do not seem very coherent either.
 * Make the otherlibraries target depend on ocamltools as is done on Unix
 * Make the ocamldebugger target depend on otherlibraries as is done on Unix
 * Since commit dd74659c6888e, the coreboot target on the Unix build
   system calls make promote with CAMLRUN=byterun/ocamlrun, which
   has never been the case on Windows. After this PR, coreboot passes
   CAMLRUN=byterun/ocamlrun to promote target on Windows, too.

The Unix and Windows versions of the core, opt, opt.opt and runtop
rules somehow differ from each other. Currently this differences have been
preserved, but all these rules should probably be unified, ultimately.

Any help in this area would be greatly appreciated.

### CODE REVIEW

**Comment by dra27** [comment]
Even before starting to test and review further - congratulations on getting to this stage, @shindere!

Mainly so we can be lazy when new commits come in, could you push a Changes entry just to keep Travis happy?

**Comment by shindere** [comment]
Hey David, very nice to read you after such a long time. :)

David Allsopp (2017/02/08 09:47 -0800):
> Even before starting to test and review further - congratulations on
> getting to this stage, @shindere!

Thanks a lot for your so prompt response and support, @dra27! It's very
encouraging and stimulating!

> Mainly so we can be lazy when new commits come in, could you push a
> Changes entry just to keep Travis happy?

Sure! Sorry for having forgotten, thanks a lot for the reminder. It's
now done.

File: configure
Comment (damiendoligez) [inline configure]: It would be cleaner to use a double-backslash here.

**Review by damiendoligez — APPROVED** [review]
Looks good.

**Comment by shindere** [comment]
Hello Damien, thanks a lot for your review,

Damien Doligez (2017/02/13 07:51 -0800):
> damiendoligez approved this pull request.
> 
> Looks good.

Great. Did you have a look to the yet-unmerged rules, though?

> > @@ -1998,9 +2008,7 @@ echo "DYNLINKOPTS=$dllib" >> Makefile
>  echo "OTHERLIBRARIES=$otherlibraries" >> Makefile
>  echo "CC_PROFILE=$cc_profile" >> Makefile
>  echo "SYSTHREAD_SUPPORT=$systhread_support" >> Makefile
> -echo "PARTIALLD=$partialld" >> Makefile
> -echo "PACKLD=\$(PARTIALLD) \$(NATIVECCLINKOPTS) -o " \
> -  | sed -e 's/ $/\\ /' >> Makefile
> +echo "PACKLD=$partialld $nativecclinkopts -o\ " >> Makefile
> 
> It would be cleaner to use a double-backslash here.

Will fix this ASAP and report here when it is done. Thanks.

**Comment by adrien-n** [comment]
Sidenote: GNU make has the -C option which chdir() to its argument. This is useful to replace the "cd foo && make ..." pattern since it can avoid spawning a shell but while I know that pmake has that optimization (i.e. looking for shell characters in the recipes and only spawning a shell when some are found), I'm not sure that GNU make also has it. If it does, that could be an easy speed win on Windows. Obviously, that's something to look at only later on.

**Review by adrien-n — COMMENTED** [review]
I would actually have installed at least README.{adoc,win32} and LICENSE on every platform instead. Definitely not at the root of the installation but somewhere under share/ or an equivalent (this seems to be meant for the OCaml Windows distribution and I haven't used it enough to know the corresponding needs).

Unfortunately it seems share/ is completely unused in the compiler but I know of no distributor that would want to skip LICENSE. I'm pretty sure all distributions copy that file manually.

This is however better handled as part of another set of changes.

**Comment by shindere** [comment]
adrien-n (2017/02/13 11:25 -0800):
> Sidenote: GNU make has the -C option which chdir() to its argument.
> This is useful to replace the "cd foo && make ..." pattern since it
> can avoid spawning a shell but while I know that pmake has that
> optimization (i.e. looking for shell characters in the recipes and
> only spawning a shell when some are found), I'm not sure that GNU make
> also has it. If it does, that could be an easy speed win on Windows.
> Obviously, that's something to look at only later on.

Yeah, the PR already makes use of -C where possible.
If you find places where it could be used but is not, please do not
hesitate to report.

File: Makefile.shared
Comment (adrien-n) [inline Makefile.shared]: I would prefer the "$< > $@" to be kept on a separate line (in order to make the addition of the WITH_SUPERDUPER_EVERYTHING option simpler).

File: Makefile.shared
Comment (adrien-n) [inline Makefile.shared]: Can you add spaces around the first '=' ?

I would also prefer to have always have spaces around makes =, := and ?= but don't mind much. This line however is a bit difficult to parse. I would also prefer the quotes to be around the whole new value rather than only $(PATH).

File: Makefile.shared
Comment (adrien-n) [inline Makefile.shared]: Same as with the previous commit, I would prefer spaces around := .

**Review by adrien-n — COMMENTED** [review]
For future reference, why are such files modified at install-time rather than during the build?

**Review by adrien-n — COMMENTED** [review]
I would install the manpages everywhere. While currently uncommon, it is possible to have a manpage reader with UNIX_OR_WIN32 = win32. This is probably already the case with msys* setups. Moreover it is simpler to remove them if they're not wanted in a specific distribution than it is to add them if they are.

File: Makefile.nt
Comment (adrien-n) [inline Makefile.nt]: I'm not fond of having half of the test as gnu make conditional and the other one as a shell one. Could you make everything either gnu make conditionals or shell conditionals?

**Review by adrien-n — COMMENTED** [review]
Could this be simplified with a rule like

asmcomp/%.ml: asmcomp/$(ARCH)/%.ml
  ln -s $(subst asmcomp/,,$<) $@

I don't remember if stems work well for such cases and I'm not going to test this tonight.

Such a change should also probably wait until after the current changes are merged.

**Comment by dbuenzli** [comment]
> I would actually have installed at least README.{adoc,win32} and LICENSE on every platform instead. Definitely not at the root of the installation but somewhere under share/ or an equivalent (this seems to be meant for the OCaml Windows distribution and I haven't used it enough to know the corresponding needs).

Just to support this further, doing so in a configurable directory would allow `opam` to install docs at the right place so that `odig` can find them. Reading these files for the end user then becomes an `odig {changes,readme,license} ocaml` invocation away.

File: Makefile.shared
Comment (adrien-n) [inline Makefile.shared]: Any news about this? Is bootstrapping even supposed to be possible on Windows?

**Comment by adrien-n** [comment]
Well, thank you and congratulations for carrying this through! I have to admit that I had troubles remaining focused throughout the review. This makes the work involved all the more palpable.

Also, now I see that my remark about make -C should have waited a bit more because you changed many, if not all, calls already.

I've refrained from writing down a few comments that are really best handled later on, especially with the runtime/coreall/core/... rules because things should work the same, these things can wait a bit more and will also be easier to change later on (it should also be much easier to get everyone involved into simplifying such lines now rather than when the differences were hidden away in a windows-specific file).

In the same spirit, if GNU make does not spawn an intermediate shell when it's not needed, several simple but common patterns should bring vast improvements, but these will also be a better fit for another batch.

Moreover, I expect these changes to be conflict-friendly and adding more changes will only make conflicts more likely and fixing horrible so I'm in favor of merging quickly.

File: Makefile
Comment (damiendoligez) [inline Makefile]: You forgot the `.PHONY` for distclean.

**Comment by damiendoligez** [comment]
@shindere 
>Did you have a look to the yet-unmerged rules, though?

I think they can wait for another PR.

**Comment by shindere** [comment]
Damien Doligez (2017/02/13 07:51 -0800):
> damiendoligez approved this pull request.
> 
> Looks good.
> 
> 
> > @@ -1998,9 +2008,7 @@ echo "DYNLINKOPTS=$dllib" >> Makefile
>  echo "OTHERLIBRARIES=$otherlibraries" >> Makefile
>  echo "CC_PROFILE=$cc_profile" >> Makefile
>  echo "SYSTHREAD_SUPPORT=$systhread_support" >> Makefile
> -echo "PARTIALLD=$partialld" >> Makefile
> -echo "PACKLD=\$(PARTIALLD) \$(NATIVECCLINKOPTS) -o " \
> -  | sed -e 's/ $/\\ /' >> Makefile
> +echo "PACKLD=$partialld $nativecclinkopts -o\ " >> Makefile
> 
> It would be cleaner to use a double-backslash here.

Fixed, thanks.

**Comment by shindere** [comment]
adrien-n (2017/02/13 11:37 -0800):
> adrien-n commented on this pull request.
> 
> I would actually have installed at least README.{adoc,win32} and LICENSE on every platform instead. Definitely not at the root of the installation but somewhere under share/ or an equivalent (this seems to be meant for the OCaml Windows distribution and I haven't used it enough to know the corresponding needs).
> 
> Unfortunately it seems share/ is completely unused in the compiler but I know of no distributor that would want to skip LICENSE. I'm pretty sure all distributions copy that file manually.
> 
> This is however better handled as part of another set of changes.

Indeed. Thanks for having raised the point, though.

**Comment by shindere** [comment]
adrien-n (2017/02/13 12:01 -0800):
> adrien-n commented on this pull request.
> 
> 
> 
> > @@ -269,8 +268,7 @@ utils/config.ml: utils/config.mlp config/Makefile
>  	    -e 's|%%TARGET%%|$(TARGET)|' \
>  	    -e 's|%%WITH_FRAME_POINTERS%%|$(WITH_FRAME_POINTERS)|' \
>  	    -e 's|%%WITH_PROFINFO%%|$(WITH_PROFINFO)|' \
> -	    -e 's|%%WITH_SPACETIME%%|$(WITH_SPACETIME)|' \
> -	    utils/config.mlp > utils/config.ml
> +	    -e 's|%%WITH_SPACETIME%%|$(WITH_SPACETIME)|' $< > $@
> 
> I would prefer the "$< > $@" to be kept on a separate line (in order
> to make the addition of the WITH_SUPERDUPER_EVERYTHING option
> simpler).

Fixed, thanks.

(although all this code is expected to disappear when switching to
autoconf)

**Comment by shindere** [comment]
adrien-n (2017/02/13 12:17 -0800):
> adrien-n commented on this pull request.
> 
> 
> 
> > +
> +INSTALL_BINDIR=$(DESTDIR)$(BINDIR)
> +INSTALL_LIBDIR=$(DESTDIR)$(LIBDIR)
> +INSTALL_COMPLIBDIR=$(DESTDIR)$(COMPLIBDIR)
> +INSTALL_STUBLIBDIR=$(DESTDIR)$(STUBLIBDIR)
> +INSTALL_MANDIR=$(DESTDIR)$(MANDIR)
> +
> +RUNTOP=./byterun/ocamlrun ./ocaml \
> +  -nostdlib -I stdlib \
> +  -noinit $(TOPFLAGS) \
> +  -I otherlibs/$(UNIXLIB)
> +NATRUNTOP=./ocamlnat$(EXE) -nostdlib -I stdlib -noinit $(TOPFLAGS)
> +ifeq "UNIX_OR_WIN32" "unix"
> +EXTRAPATH=
> +else
> +EXTRAPATH=PATH=otherlibs/win32unix:"$(PATH)"
> 
> Can you add spaces around the first '=' ?

Sure, done.

> I would also prefer to have always have spaces around makes =, := and
> ?= but don't mind much.

That's a good idea, yes. Didn't check at the moment but will do in
further commits.

> This line however is a bit difficult to parse.

Indeed.

> I would also prefer the quotes to be around the whole new value rather
> than only $(PATH).

Done

adrien-n (2017/02/13 12:19 -0800):
> adrien-n commented on this pull request.
> 
> 
> 
> > @@ -253,6 +253,18 @@ else
>  EXTRAPATH=PATH=otherlibs/win32unix:"$(PATH)"
>  endif
>  
> +BOOT_FLEXLINK_CMD=
> +
> +ifeq "$(UNIX_OR_WIN32)" "win32"
> +FLEXDLL_SUBMODULE_PRESENT:=$(wildcard flexdll/Makefile)
> +ifeq "$(FLEXDLL_SUBMODULE_PRESENT)" ""
> +  BOOT_FLEXLINK_CMD=
> +else
> +  BOOT_FLEXLINK_CMD=FLEXLINK_CMD="../boot/ocamlrun ../flexdll/flexlink.exe"
> +  CAMLOPT:=OCAML_FLEXLINK="boot/ocamlrun flexdll/flexlink.exe" $(CAMLOPT)
> 
> Same as with the previous commit, I would prefer spaces around := .

Done, too.

**Comment by shindere** [comment]
adrien-n (2017/02/13 12:47 -0800):
> adrien-n commented on this pull request.
> 
> I would install the manpages everywhere. While currently uncommon, it
> is possible to have a manpage reader with UNIX_OR_WIN32 = win32. This
> is probably already the case with msys* setups. Moreover it is simpler
> to remove them if they're not wanted in a specific distribution than
> it is to add them if they are.

The non-installation had been discussed on the caml-devel mailing-list,
but I guess we could provide a dedicated make target, for instance.

**Comment by shindere** [comment]
adrien-n (2017/02/13 12:52 -0800):
> adrien-n commented on this pull request.
> 
> 
> 
> > @@ -180,9 +180,11 @@ endif
>  	if test -n "$(WITH_DEBUGGER)"; then \
>  	  $(MAKE) -C debugger install; \
>  	fi
> +ifeq "$(UNIX_OR_WIN32)" "win32"
>  	if test -n "$(FLEXDLL_SUBMODULE_PRESENT)"; then \
> 
> I'm not fond of having half of the test as gnu make conditional and
> the other one as a shell one. Could you make everything either gnu
> make conditionals or shell conditionals?

I think I'll change this code in later PRs anyway so I'd rather leave it
as it is for the moment.

**Comment by shindere** [comment]
adrien-n (2017/02/13 13:09 -0800):
> adrien-n commented on this pull request.
> 
> Could this be simplified with a rule like
> 
> asmcomp/%.ml: asmcomp/$(ARCH)/%.ml
>   ln -s $(subst asmcomp/,,$<) $@
> 
> I don't remember if stems work well for such cases and I'm not going
> to test this tonight.

I tried but didn't find an obvious way so I left it for later.

> Such a change should also probably wait until after the current
> changes are merged.

Yes, and there may be a better way to refactor all this part.

**Comment by shindere** [comment]
adrien-n (2017/02/13 13:29 -0800):
> adrien-n commented on this pull request.
> 
> 
> 
> > @@ -347,6 +347,22 @@ coldstart:
>  	cd stdlib; cp $(LIBFILES) ../boot
>  	cd boot; $(LN) ../byterun/libcamlrun.$(A) .
>  
> +# Recompile the core system using the bootstrap compiler
> +.PHONY: coreall
> +coreall:
> +	$(MAKE) ocamlc
> +	$(MAKE) ocamllex ocamlyacc ocamltools library
> +
> +# Build the core system: the minimum needed to make depend and bootstrap
> +.PHONY: core
> +core:
> +ifeq "$(UNIX_OR_WIN32)" "unix"
> +	$(MAKE) coldstart
> +else # Windows, to be fixed!
> 
> Any news about this?

We will finish the merging of the yet unmerged rules after the beta
release, I think.

> Is bootstrapping even supposed to be possible on Windows?

I believe so.

**Comment by shindere** [comment]
adrien-n (2017/02/13 13:54 -0800):
> Well, thank you and congratulations for carrying this through!

And thanks to you for having patiently reviewed all this! The Changes
entry for this PR has been updated to mention this.

> I've refrained from writing down a few comments that are really best
> handled later on, especially with the runtime/coreall/core/... rules
> because things should work the same, these things can wait a bit more
> and will also be easier to change later on (it should also be much
> easier to get everyone involved into simplifying such lines now rather
> than when the differences were hidden away in a windows-specific
> file).

Please do not hesitate to share all the ideas you may have, let's make
sure they do not get lost!

> In the same spirit, if GNU make does not spawn an intermediate shell
> when it's not needed, several simple but common patterns should bring
> vast improvements, but these will also be a better fit for another
> batch.

What do you have in mind, here?

> Moreover, I expect these changes to be conflict-friendly and adding
> more changes will only make conflicts more likely and fixing horrible
> so I'm in favor of merging quickly.

Thanks! Today hopefully! :)

**Comment by shindere** [comment]
Damien Doligez (2017/02/14 06:17 -0800):
> You forgot the `.PHONY` for distclean.

Oops, fixed, thanks.

**Pull request merged**
